firebase-firestore/src/main/java/com/google/firebase/firestore/core/FieldFilter.java

```diff
@@ -15,37 +15,81 @@
 package com.google.firebase.firestore.core;
 
 import static com.google.firebase.firestore.util.Assert.hardAssert;
+import static com.google.firebase.firestore.util.Preconditions.checkNotNull;
 
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import com.google.firebase.firestore.Filter;
 import com.google.firebase.firestore.model.Document;
 import com.google.firebase.firestore.model.FieldPath;
 import com.google.firebase.firestore.model.Values;
 import com.google.firebase.firestore.util.Assert;
+import com.google.firestore.v1.StructuredQuery;
 import com.google.firestore.v1.Value;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.List;
 
-/** Represents a filter to be applied to query. */
+/** Represents a single-field filter to be applied to query. */
 public class FieldFilter extends Filter {
+  public enum Operator {
+    LESS_THAN("<"),
+    LESS_THAN_OR_EQUAL("<="),
+    EQUAL("=="),
+    NOT_EQUAL("!="),
+    GREATER_THAN(">"),
+    GREATER_THAN_OR_EQUAL(">="),
+    ARRAY_CONTAINS("array_contains"),
+    ARRAY_CONTAINS_ANY("array_contains_any"),
+    IN("in"),
+    NOT_IN("not_in");
+
+    private final String text;
+
+    Operator(String text) {
+      this.text = text;
+    }
+
+    @Override
+    public String toString() {
+      return text;
+    }
+  }
+
   private final Operator operator;
 
   private final Value value;
 
   private final FieldPath field;
 
+  private final Object valueObject;
+
   /**
    * Creates a new filter that compares fields and values. Only intended to be called from
    * Filter.create().
    */
+  protected FieldFilter(FieldPath field, Operator operator, Object valueObject) {
+    this.field = field;
+    this.operator = operator;
+    this.valueObject = valueObject;
+    this.value = null;
+  }
+
+  /**
+   * Creates a new filter that compares fields and values. Only intended to be called from
+   * Filter.create(). This is a more specialized constructor that takes a parsed Value.
+   */
   protected FieldFilter(FieldPath field, Operator operator, Value value) {
     this.field = field;
     this.operator = operator;
     this.value = value;
+    this.valueObject = null;
```

Comment:
As stated elsewhere, I hope that we can refactor this so the `valueObject` stays in the API layer.

Reply:
Acknowledged.